From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: kernel: do full redraw of the 3270 screen on reconnect
Patch-mainline: Queued in linux/linux-next
Git-commit: 05bfd70bcdd3cd12c061cb77b73a11ba6f87379d
References: bnc#943476, LTC#129509

Description:  kernel: do full redraw of the 3270 screen on reconnect
Symptom:      A 3270 terminal is not redrawn correctly if the device
              is disconnect and then reconnect again. Only after a
              clear screen the correct output is display again.
Problem:      On reconnect of a 3270 device a unsolicited device end
              interrupt is made pending. This interrupt is ignored by
              the 3270 device driver.
Solution:     Do a full redraw of the 3270 screen after a unsolicited
              device end has been received.
Reproduction: Disconnect and reconnect a 3270 terminal.

Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/char/con3270.c |    4 ++++
 drivers/s390/char/tty3270.c |    4 ++++
 2 files changed, 8 insertions(+)

--- a/drivers/s390/char/con3270.c
+++ b/drivers/s390/char/con3270.c
@@ -413,6 +413,10 @@ con3270_irq(struct con3270 *cp, struct r
 		else
 			/* Normal end. Copy residual count. */
 			rq->rescnt = irb->scsw.cmd.count;
+	} else if (irb->scsw.cmd.dstat & DEV_STAT_DEV_END) {
+		/* Interrupt without an outstanding request -> update all */
+		cp->update_flags = CON_UPDATE_ALL;
+		con3270_set_timer(cp, 1);
 	}
 	return RAW3270_IO_DONE;
 }
--- a/drivers/s390/char/tty3270.c
+++ b/drivers/s390/char/tty3270.c
@@ -662,6 +662,10 @@ tty3270_irq(struct tty3270 *tp, struct r
 		else
 			/* Normal end. Copy residual count. */
 			rq->rescnt = irb->scsw.cmd.count;
+	} else if (irb->scsw.cmd.dstat & DEV_STAT_DEV_END) {
+		/* Interrupt without an outstanding request -> update all */
+		tp->update_flags = TTY_UPDATE_ALL;
+		tty3270_set_timer(tp, 1);
 	}
 	return RAW3270_IO_DONE;
 }
